lib/IDE/CodeCompletionResultType.cpp

```diff
@@ -11,50 +11,227 @@
+    auto UnwrappedOptionalType = possibleTy->getOptionalObjectType();
+    while (UnwrappedOptionalType) {
+      ContextualTypes.emplace_back(
+          USRBasedType::fromType(UnwrappedOptionalType, Arena),
+          /*IsConvertible=*/true);
+      UnwrappedOptionalType = UnwrappedOptionalType->getOptionalObjectType();
+    }
+
+    // If the contextual type is an opaque return type, make the protocol a
+    // contextual type. E.g. if we have
+    //   func foo() -> some View { #^COMPLETE^# }
+    // we should show items conforming to `View` as convertible.
+    if (auto OpaqueType = possibleTy->getAs<ArchetypeType>()) {
+      llvm::SmallVector<const USRBasedType *, 1> USRConforms;
+      for (auto Proto : OpaqueType->getConformsTo()) {
+        USRConforms.push_back(
+            USRBasedType::fromType(Proto->getDeclaredInterfaceType(), Arena));
+      }
+      // Archetypes are also be used to model generic return types, in which
+      // case they don't have any conformsTo entries. We simply ignore those.
+      if (!USRConforms.empty()) {
+        ContextualTypes.emplace_back(USRConforms, /*IsConvertible=*/false);
+      }
+    }
+  }
+}
+
+TypeRelation
+USRBasedTypeContext::typeRelation(const USRBasedType *ResultType) const {
+  if (ResultType == this->VoidType) {
+    // Void is not convertible to anything and we don't report Void <-> Void
+    // identical matches (see USRBasedType::typeRelation). So we don't have to
+    // check anything if the result returns Void.
+    return TypeRelation::Unknown;
+  }
+
+  TypeRelation Res = TypeRelation::Unknown;
+  for (auto &ContextualType : ContextualTypes) {
+    Res = std::max(Res, ContextualType.typeRelation(ResultType, VoidType));
+  }
+  return Res;
+}
+
+// MARK: - USRBasedTypeArena
+
+const USRBasedType *USRBasedTypeArena::getVoidType(const ASTContext &ASTCtx) {
+  if (!VoidType) {
+    VoidType = USRBasedType::fromType(ASTCtx.getVoidType(), *this);
+  }
+  return VoidType;
+}
+
+// MARK: - USRBasedType
+
+const USRBasedType *USRBasedType::null(USRBasedTypeArena &Arena) {
+  return USRBasedType::fromUSR(/*USR=*/"", /*Supertypes=*/{}, Arena);
+}
+
+const USRBasedType *
+USRBasedType::fromUSR(StringRef USR, ArrayRef<const USRBasedType *> Supertypes,
+                      USRBasedTypeArena &Arena) {
+  auto ExistingTypeIt = Arena.CanonicalTypes.find(USR);
+  if (ExistingTypeIt != Arena.CanonicalTypes.end()) {
+    assert(ArrayRef<const USRBasedType *>(ExistingTypeIt->second->Supertypes) ==
+               Supertypes &&
+           "Same USR but different supertypes?");
+    return ExistingTypeIt->second;
+  }
+  const USRBasedType *Result =
+      new (Arena.Allocator) USRBasedType(USR, Supertypes);
+  Arena.CanonicalTypes[USR] = Result;
+  return Result;
+}
+
+const USRBasedType *USRBasedType::fromType(Type Ty, USRBasedTypeArena &Arena) {
+  // FIXME: We can't mangle archetypes, so we treat them as null USRBasedTypes.
```

Comment:
This actually works. We are reporting convertible type relations for types that conform to `MyProto` here. I added a test case for it.

Reply:
I mean when suggesting it. The test case would be something like:
```swift
//-- Lib.swift
func returnSomeMyProto() -> some MyProto { ... }

//-- test.swift
import Lib
func foo<T: MyProto>(_: T)
func test() {
  foo(#^HERE^#)
} 
```

Reply:
I see, I decided to special case `some` with a single protocol for now as representing the conjunction of two types isn’t currently supported in `USRBasedType` and it’s not too simple to put it in. Since the entire USRBasedType thing is just an under-approximation anyway, I think it’s fine if we can’t represent `some MyProto & MyOtherProto` for now. https://github.com/apple/swift/pull/41000/commits/6aff8ef500ee97e10ed55b84a8da5ac298391655